### ISSUE TITLE
date and tz - 2 libraries in one - sure this is the path to go?

### ISSUE THREAD

**Issue by daixtrose** [HowardHinnant/date#228]
While reworking and reorganizing the recently proposed changes in #210 in order to resubmit PRs in smaller portions I took a closer look at the docs. My aim still is to provide best-practice CMake support. IIUC we have the following situation:
- There exists a header-only library `date`, completely residing in `include/date`
  - This could be handled easily by CMake by following [advice for header only libraries](http://www.mariobadr.com/creating-a-header-only-library-with-cmake.html)  
- There exists a header/source mixed library `tz` consisting of  `include/date/tz.h`, `include/date/tz.cpp` and `src/tz.cpp`
  - This has to be handled as already started in #210 creating directives for [building/installing](https://github.com/ReinventorsOfTheWheel/date/blob/d43dcc83d3e79f8d95994a2f9de09100abedf607/CMakeLists.txt) and/or [packaging](https://github.com/ReinventorsOfTheWheel/date/blob/d43dcc83d3e79f8d95994a2f9de09100abedf607/CMakeCPack.cmake)  

My question to @HowardHinnant: In the long term do you plan to keep these 2 distinct libraries together or move those into separate repositories? Note that I am completely unbiased and neither in favor of one or the other solution. I am just asking because this might influence the CMake structure.

I can bundle both libraries into one called `date` or try to configure CMake in such a way that both libraries (`date` and `tz`) can be imported separately while still living below the same directory structure.

Even in the latter case I would prefer all `tz`stuff to be in its own directories, i.e.
```bash
mkdir include include/tz
git mv include/date/tz* include/tz
``` 
What do you think?

As a sidenote: IMHO the library names are too short, therefore at risk to match other library names. For `date` I would consider a rename to something more unique. Otherwise the recent directory structure reorganization is only half of the cure. I also propose to find a name that can be easily found in a search engine.

**Comment by HowardHinnant** [comment]
Thanks for working on this.

I have quite a few existing clients using this library.  Changing the directory structure once was a big deal.  I didn't do it lightly as I knew I was breaking a lot of code.  It is now set in stone unless a seriously evil bug is revealed that is caused by the current structure and would be fixed by a new structure.

My best guess is that date.h alone could be ignored by CMake support.  It is a single-header library.  People don't need CMake to use it, except maybe for an include search path to date/include.  I recommend concentrating on tz.h.

The long term plan is to move both of these libraries into the std `<chrono>` header (https://wg21.link/p0355), though success in such an endeavor is far from assured.

**Comment by daixtrose** [comment]
I understand. So for the CMake part do you want both libraries "packed" into one library called `date` or do you prefer a CMake file advertizing both separately?

**Comment by HowardHinnant** [comment]
I recommend just doing `tz.h`/`tz.cpp` for CMake.  Once you have those, you implicitly get all the functionality in `date.h` anyway.  And if you just want `date.h` without support or expense of time zones, CMake is really overkill anyway.

For those few who only want `date.h`, but don't want to go to the trouble of adding an include path to their own build systems, let them get `date.h` by using the tz lib.

I also would not include the tests in CMake.  There's already a shell script to run the tests.  And the tests are designed to fail on non-conforming platforms, and most are.  That will just frustrate people.

CMake could perhaps make it easier to deal with the installation options documented at https://howardhinnant.github.io/date/tz.html#Installation .  Perhaps it could even auto-install curl and 7-zip on Windows (I have no idea).  I hear that dealing with the installation of tz.h is the real pain point.

**Comment by daixtrose** [comment]
I disagree on the "CMake for date.h is overkill". If you are in a CMake context, you are happy to see each library play well with the others. The reason I went into this is that I wanted to use `date` in another project that is driven by CMake. I wanted to make sure the library was included via git submodules in order to obtain the latest and greatest version available whenever I perform an update. For this use case a one-liner like e.g. `target_link_library(my-own-lib date)` is much better than adding include directories by hand, because the information in a library specific `CMakeLists.txt` propagates up the whole tree. This starts to pay off when I include `date` in several projects. 

I agree with the installation to be handled with CMake. [DownloadProject](https://github.com/Crascit/DownloadProject) will be really useful for getting [libcurl](https://github.com/curl/curl) and all other dependencies. Challenge accepted.   

I disagree with the tests requiring `/bin/sh` even on windows. Also it makes it hard to test with different compilers, etc. OTOH I understand that this has no high priority for you and therefore will postpone this part.

**Comment by HowardHinnant** [comment]
Ok, thanks for the feedback.  I'm a CMake neophyte and have much to learn.  Indeed, my plan is to use your work to educate myself in this area. :-)

Is there a convenient way to tell CMake to just search date/include, and not try to compile date/src/tz.cpp or install curl or 7-zip?  If so, that's all we need for date.h.

**Comment by daixtrose** [comment]
Yes. There Is a way. I will create a proposal.

**Comment by MikeGitb** [comment]
Just wanted to say that I'm very interested in full cmake support with separate library targets for date and tz. I hacked a little on it while trying to improve the integration in vcpkg (e.g. automatically downloading the Timezone database) but then got stuck with porting the tests and also wanted to wait until there was an official release with the new directory structure.

**Comment by Erroneous1** [comment]
This is just my two cents. I have a project that uses this library in cmake. It provides 2 "libraries" date and date-tz. If I just need date I can "link" to it, or I can link to date-tz if I need timezone support. Here's the important bits:

```cmake
add_library(date INTERFACE)
target_include_directories(date INTERFACE include)
add_library(date-tz STATIC src/tz.cpp)
target_link_libraries(date-tz date)
```

**Comment by daixtrose** [comment]
Thanks for your comments. Please forgive my delay. I have something nearly ready, but still want to fix some details. Stay tuned.

**Comment by anupamsr** [comment]
Sorry to butt in here but I am using date/tz both using a git submodule. For that purpose I have hacked together a CMakeLists.txt that compiles tz library (static and shared), and my master CMakeLists.txt copies this over the date repo (after cloning it) so it all works nicely. The only problem is that git will perpetually compain that date repo has become 'dirty'. Hope that makes sense. Here is what I got:

[CMakeLists.txt](https://github.com/HowardHinnant/date/files/1578916/CMakeLists.txt)

And in master CMakeLists.txt
```
 # Find date.h/tz.h
 set(DATE_TZ_DIR "external/date")
 find_path(DATE_TZ_INCLUDE_DIR "date/date.h")
 if((NOT DATE_TZ_INCLUDE_DIR) OR (NOT EXISTS ${DATE_TZ_INCLUDE_DIR}))
     message(STATUS "Unable to find date.h, using local copy...")
     execute_process(COMMAND git submodule update --init -- external/date WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
     execute_process(COMMAND ${CMAKE_COMMAND} -E copy
         ${CMAKE_SOURCE_DIR}/etc/cmake/date/CMakeLists.txt
         ${CMAKE_SOURCE_DIR}/${DATE_TZ_DIR})
     set(DATE_TZ_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/${DATE_TZ_DIR}/include")
     add_subdirectory(${DATE_TZ_DIR})
 endif()
 include_directories(${DATE_TZ_INCLUDE_DIR})
```

Eagerly waiting for your solution @daixtrose
